### PR TITLE
Fixes #15696 - Added memory check to installer

### PR DESCRIPTION
**Pull request by chris1984**
Added memory check to installer to make sure system has decent ram before running the installer.

### CODE REVIEW

**Comment by iNecas** [comment]
There should be still a way at least to force skip this check: As a developer, I need to run more than one VM sometimes, and I can take the responsibility for the memory allocation and that it might not be enough for some cases.

**Comment by stbenjam** [comment]
Unfortunately, `boot` (where we add options) runs after anything in checks.  We'll have to be a little creative.  Either read an environment variable, or use the option anyway with a little creativity (check something like `ARGV.incldue?('--disable-memory-check')` instead of using the app_value helper).

**Comment by sideangleside** [comment]
Also, would it be a better idea to use facter to get the amount of installed memory? That way you don't have to parse the output of the `free` command. 

~~~
begin
  require 'rubygems'
rescue LoadError
end
require 'facter'

total_ram = Facter.value(:memorysize_mb)
~~~

**Comment by chris1984** [comment]
@stbenjam updated with a env variable check here is my output from my testing

**Comment by chris1984** [comment]
without variable

[root@centos7-katello-3-1 ~]# foreman-installer --scenario katello -v
This system has less than 8 GB of total memory. Please have at least 8 GB of total ram free before running the installer.
Your system does not meet configuration criteria
[ INFO 2016-08-25 11:47:54 verbose] Executing hooks in group pre_migrations
[ INFO 2016-08-25 11:47:54 verbose] All hooks in group pre_migrations finished


With variable:

[root@centos7-katello-3-1 ~]# export DISABLE_INSTALLER_CHECKS="true"
[root@centos7-katello-3-1 ~]# echo $DISABLE_INSTALLER_CHECKS
true

[root@centos7-katello-3-1 ~]# foreman-installer --scenario katello -v
DISABLEDCHECK
[ INFO 2016-08-25 11:48:49 verbose] Executing hooks in group pre_migrations
[ INFO 2016-08-25 11:48:49 verbose] All hooks in group pre_migrations finished
[ INFO 2016-08-25 11:48:49 verbose] Executing hooks in group boot
[ INFO 2016-08-25 11:48:49 verbose] All hooks in group boot finished
[ INFO 2016-08-25 11:48:49 verbose] Executing hooks in group init
[ INFO 2016-08-25 11:48:49 verbose] All hooks in group init finished
[ INFO 2016-08-25 11:50:25 verbose] Executing hooks in group post
  Success!
  * Katello is running at https://centos7-katello-3-1.example.com
  * To install additional capsule on separate machine continue by running:

      capsule-certs-generate --capsule-fqdn "$CAPSULE" --certs-tar "~/$CAPSULE-certs.tar"

  The full log is at /var/log/foreman-installer/katello.log
[ INFO 2016-08-25 11:50:25 verbose] All hooks in group post finished

**Comment by chris1984** [comment]
@sideangleside updated to use facter :)

**Comment by stbenjam** [comment]
We don't use environment variables anywhere else that I know of, I think it's better if you use installer params to bypass the check, plenty of examples in the boot directory, for example;

https://github.com/Katello/katello-installer/blob/master/hooks/boot/30-upgrade.rb

And you check the value with `app_value(:upgrade)`

File: checks/memory.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# Check for the presence of the DISABLE_INSTALLER_CHECKS
+# Environment Variable.  If it is nil, we dont want
+# to disable checking. Otherwise TRUE = nochecks
+# FALSE (or any other value) = check
+
+begin
+  require 'rubygems'
+rescue LoadError
+end
+require 'facter'
```
Comment (stbenjam) [inline checks/memory.rb]: You cannot use the facter gem directly, this is what we were talking about on IRC.  In Puppet 4, the gem isn't accessible to the installer.  You would need to shell out and call facter binary or use the original version you had.

File: hooks/pre/17-memory_check.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+min_ram = 8388608
+       
+# call mem_check if flag is called
+if app_value(:disable_system_checks) then
+logger.warn 'Skipping system checks.'
+else
+ if min_ram > total_ram
+logger.error 'This system has less than 8 GB of total memory. Please have at least 8 GB of total ram free before running the installer.'
+kafo.class.exit 1
+ end
```
Comment (stbenjam) [inline hooks/pre/17-memory_check.rb]: Please fix indentation, all ruby code needs to follow [our style guide](http://www.katello.org/developers/style/ruby.html)

File: hooks/boot/21-disable_system_checks.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Add reset option
+app_option(
+  '--disable_system_checks',
```
Comment (stbenjam) [inline hooks/boot/21-disable_system_checks.rb]: Please use dashes not underscores so it's consistent with all other installer options

**Comment by stbenjam** [comment]
Looks great!! One tiny nitpick, can you make the number prefixes of the file in `boot/` and the file in `pre/` the same? The rest all match up.

**Comment by stbenjam** [comment]
Thanks, looks great, ACK!

**Comment by chris1984** [comment]
@stbenjam can I get an updated check when you get a free moment

**Pull request merged**